gix-blame/src/types.rs

```diff
@@ -81,60 +101,51 @@ impl BlameRanges {
     /// The range should be 1-based inclusive.
     /// If the new range overlaps with or is adjacent to an existing range,
     /// they will be merged into a single range.
-    pub fn add_range(&mut self, new_range: RangeInclusive<u32>) {
-        self.merge_range(new_range);
+    pub fn add_range(&mut self, new_range: RangeInclusive<u32>) -> Result<(), Error> {
+        match self {
+            Self::PartialFile(_) => {
+                let zero_based_range = Self::inclusive_to_zero_based_exclusive(new_range);
+                self.merge_range(zero_based_range)
+            }
+            _ => Err(Error::InvalidOneBasedLineRange),
+        }
     }
 
     /// Attempts to merge the new range with any existing ranges.
     /// If no merge is possible, add it as a new range.
-    fn merge_range(&mut self, new_range: RangeInclusive<u32>) {
-        // Check if this range can be merged with any existing range
-        for range in &mut self.ranges {
-            // Check if ranges overlap or are adjacent
-            if new_range.start() <= range.end() && range.start() <= new_range.end() {
-                *range = *range.start().min(new_range.start())..=*range.end().max(new_range.end());
-                return;
+    fn merge_range(&mut self, new_range: Range<u32>) -> Result<(), Error> {
+        match self {
+            Self::PartialFile(ref mut ranges) => {
+                // Check if this range can be merged with any existing range
+                for range in &mut *ranges {
+                    // Check if ranges overlap
+                    if new_range.start <= range.end && range.start <= new_range.end {
+                        *range = range.start.min(new_range.start)..range.end.max(new_range.end);
+                        return Ok(());
+                    }
+                    // Check if ranges are adjacent
+                    if new_range.start == range.end || range.start == new_range.end {
+                        *range = range.start.min(new_range.start)..range.end.max(new_range.end);
+                        return Ok(());
+                    }
+                }
```

Comment:
Currently, this does not prevent incorrect use in this edge case: `gix blame -L 1,3 -L 5,7 -L 2,6`. Maybe this can be turned into a `.fold` call?